### DISCUSSION TITLE
How to permit an Agent to read a file uploaded to OpenAI storage by BrowserAgent?

### DISCUSSION THREAD

**Discussion by simoninnyc** [VRSEN/agency-swarm#178] [Q&A]
I'm struggling to work out how to grant an agent the right permission to be able to read a file that another Agent has created. In my case one Agent asks the BrowserAgent to export HTML using a custom tool I created and save it in OpenAI storage. The Browser Agent confirms the file ID back to the requesting Agent and I see the file has been created when I browse to my OpenAI project storage. However my agent then reports that there was an issue accessing the HTML file. Do I need a custom tool in my Agent to retrieve the file using the file ID, or is this simply a matter of configuring the Agent correctly? Thanks for any suggestions you have. Sorry if I've missed something obvious!

**Comment by nicko-ai** [comment]
Generally, all you need to do is enable CodeInterpreter and add it to the list of tools.
Let me know if it works for you.

**Reply by simoninnyc** [reply]
Hi @bonk1t ! Thanks for your swift reply. I tried adding the CodeInterpreter to the list of tools for the Agent that needs to read the file but I'm still getting an error back "It seems there was an issue accessing the HTML file."

Here's the code for my custom tool that I've added to the BrowsingAgent:

`class ExportHtml(BaseTool):
    """This tool extracts the complete HTML source for a web page into a file and returns its file_id. You can then send this file id back to the user for further processing."""

    def run(self):
        wd = get_web_driver()
        from agency_swarm import get_openai_client
        client = get_openai_client()
        page = wd.page_source.encode('utf-8')
        
        # Save the HTML to a file
        with open("exported_file.html", "wb") as f:
            f.write(page)
        file_id = client.files.create(file=open("exported_file.html", "rb"), purpose="assistants",).id
        self._shared_state.set("file_id", file_id)
        return "Success. File exported with id: `" + file_id + "` You can now send this file id back to the user."

and the Agent that needs to read the file:

from agency_swarm.agents import Agent
from agency_swarm.tools import CodeInterpreter
class HtmlExpert(Agent):
    def __init__(self):
        super().__init__(
            name="HtmlExpert",
            description="Analyzes HTML",
            instructions="./instructions.md",
            files_folder="./files",
            schemas_folder="./schemas",
            tools=[CodeInterpreter],
            tools_folder="./tools",
            temperature=0,
            max_prompt_tokens=50000
        )

**Reply by nicko-ai** [reply]
your issue was resolved here: https://github.com/VRSEN/agency-swarm/issues/161

**Reply by simoninnyc** [reply]
Thanks for your continued help @bonk1t , I'm not sure how #161 addresses my issue as the file is being created by another Agent and not uploaded manually by the user. My intended use case is:

1. HTMLExpert requests BrowsingAgent to visit a URL and save the HTML to a file that it stores in the thread.
2. HTMLExpert examines the HTML and extracts specific values.

I can get step 1 to work fine and see the HTML file has been saved to OpenAI storage but then the HTMLExpert is unable to access the file.

**Reply by simoninnyc** [reply]
Actually I think I need to reread #161 again...

**Reply by simoninnyc** [reply]
Unfortunately I wasn't able to get the solution for #161 to work. The Send button in the Gradio chat no longer works when I implement the ExtendedAgency. I tried experimenting with the myfiles_browser tool which other Agents including the BrowserAgent seem to use but whatever I try I can't get my agent to read files that are being stored locally or within the openai  directories.

**Reply by nicko-ai** [reply]
Thank you for sharing your use case:
**- HTMLExpert requests BrowsingAgent to visit a URL and save the HTML to a file stored in the thread.**
**- HTMLExpert then examines the HTML and extracts specific values.**

Have you considered using local storage? We often use this approach, as agents can access file names through `shared_state`.

Alternatively, if you're working entirely in the cloud, you could save the files to Google Drive and return the `file_id` from the tool. This would allow you to load the file in another tool.

Please keep in mind that `shared_state` won't work if you deploy tools separately as serverless functions.

Let me know if this helps or if you need further clarification!

Make to join our Discord for faster responses and community support!

**Reply by simoninnyc** [reply]
Thanks @bonk1t , I just tried using local storage again, using the file-id as the local filename without any extension. I had to modify the FileReader stripping out the /mnt/data prefix which seems to get applied automatically:

def run(self):
        # read file
        filename = self.file_path.removeprefix('/mnt/data/')
        with open(os.path.join(self._shared_state.get("assets_dir", "./assets"), filename), "r") as f:
            file_contents = f.readlines()
        # return file contents
        return "\n".join([f"{i + 1}. {line}" for i, line in enumerate(file_contents)])

I'm now hitting a file size limit for the tool since the HTML files may be large:
"Invalid 'tool_outputs[1].output': string too long. Expected a string with maximum length 1048576, but got a string with length 1185424 instead."

I'm thinking I'll run into the same constraint even if I save the files to Google Drive, since the tool still needs to read them. Perhaps I need Vector Storage for this use case...

**Reply by nicko-ai** [reply]
At our agency, we take extra care with tool outputs to ensure they are clean and efficient. We typically don’t output file contents directly, but instead save them within the tool and return the filename/ID for reference.

If you’d like quicker assistance or would love to connect with our community, please feel free to join our Discord server! You can post any questions in the #issues-and-discussions section, and we’ll be happy to help.

Looking forward to seeing you there!

Let me know if you’d like further adjustments!
